### PR TITLE
feat: add endpoint to check if a collection exists by name

### LINKED ISSUE #3332: API for checking collection exists
**Is your feature request related to a problem? Please describe.**

Currently, there are 2 ways of checking if collection exists:

- list all collections and find name in the list
- query specific collection info and check if it is 404 or 200 response code

But in some clients if would be much more convenient to have a dedicated API which would return `true` or `false` without error http code.

**Describe the solution you'd like**

API `/collections/<collection-name>/check` and similar in gRPC

API should return `true` if collection exists and `false` if not (with all the usual result wrappers, of course)

**Describe alternatives you've considered**

Make more sophisticated and extendable API, which could return some additional information as well (but not as much as collection info)

Or make an API for batch request of collection info, so that you could specify multiple names and get results for all of them

**Additional context**

API should be consistent with our current set of endpoints

---

Note for contributors: Please consider this as tracking issue. If you think that it would be beneficial to split the task into multiple smaller PRs, please you are welcome to do so. Bounty will be rewarded for each PR independently

### PR DESCRIPTION
**Pull request by rajdip-b**
### All Submissions:

* [x] Contributions should target the `dev` branch. Did you create your branch from `dev`?
* [x] Have you followed the guidelines in our Contributing document?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../../pulls) for the same update/change?

### New Feature Submissions:

1. [x] Does your submission pass tests?
2. [x] Have you formatted your code locally using `cargo +nightly fmt --all` command prior to submission?
3. [x] Have you checked your code using `cargo clippy --all --all-features` command?

### Changes to Core Features:

* [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] Have you written new tests for your core changes, as applicable?
* [x] Have you successfully ran tests with your changes locally?

Fixes #3332
/claim #3332

### CODE REVIEW

**Comment by rajdip-b** [comment]
@generall can i get a review on this?

**Comment by generall** [comment]
> @generall can i get a review on this?

Sure! There are a few things missing:

- OpenAPI definitions for the new API
- gRPC version
- Tests

**Comment by rajdip-b** [comment]
> > @generall can i get a review on this?
> 
> Sure! There are a few things missing:
> 
>     * OpenAPI definitions for the new API
> 
>     * gRPC version
> 
>     * Tests

cool! I will get started with them!

**Comment by rajdip-b** [comment]
@generall i have added the following:
- grpc function
- openapi docs

can you please tell me where i can find the tests? I am a bit confused about it :/

**Comment by generall** [comment]
API tests usually go here: https://github.com/qdrant/qdrant/tree/master/openapi/tests/openapi_integration

**Comment by rajdip-b** [comment]
@generall can you give it a review now?

**Comment by rajdip-b** [comment]
okei hold up i feel there's something really wrong going on.

**Comment by coszio** [comment]
Code in `qdrant.rs` is mostly auto-generated from the `.proto` files. What you need to modify for gRPC interface are these `.proto` files, the auto-generation will be triggered when compiling.

**Comment by rajdip-b** [comment]
> Code in `qdrant.rs` is mostly auto-generated from the `.proto` files. What you need to modify for gRPC interface are these `.proto` files, the auto-generation will be triggered when compiling.

Gotcha!

**Comment by rajdip-b** [comment]
Okay, I have updated the `collections.proto` and `collections_service.proto` files and made the `collection_api.rs` file to do the actual implementation. Now, how do I auto generate the rust config for the proto files?

**Comment by rajdip-b** [comment]
@coszio hey, any updates?

File: src/common/collections.rs
Comment (amiremohamadi) [inline src/common/collections.rs:57]: here if any errors happen inside `get_collection`, we return `false` which is not the desired result. it's better to expose the error.

Comment (rajdip-b) [reply src/common/collections.rs:57]: Sure, I'll get this fixed

**Comment by rajdip-b** [comment]
@amiremohamadi hey man, I'm facing some build errors so have opened a thread in discord. Would you please help me a bit on this one?

**Comment by rajdip-b** [comment]
@amiremohamadi @generall can i get a review on this?

**Comment by rajdip-b** [comment]
alright so i found this bug in my code. 

```
message CheckCollectionExistsRequest {
  string collection_name = 1; // Name of the collection
}
```

This is in my `collections.proto` file. Now when i build it, it generates this:

```
#[derive(serde::Serialize)]
#[allow(clippy::derive_partial_eq_without_eq)]
#[derive(Clone, PartialEq, ::prost::Message)]
pub struct CheckCollectionExistsRequest {
    /// Name of the collection
    #[prost(string, tag = "1")]
    pub collection_name: ::prost::alloc::string::String,
}
```

I noticed that there isnt the `#[derive(validator::Validate)]` trait, which is basically causing the errors. editing qdrant.rs file isnt the way out, so im confused as to what im doing wrong in here

**Comment by coszio** [comment]
hint: check out `lib/api/build.rs`

**Comment by rajdip-b** [comment]
Yeah man I figured :) Didn't know how GRPC fits in so had to do some digging into the docs and now its fixed. Im running the tests now (finally after fixing all the breaking stuff) and updating it asap

**Comment by rajdip-b** [comment]
@coszio can you approve the workflow once?

**Comment by rajdip-b** [comment]
Fixing the tests, one sec

**Comment by rajdip-b** [comment]
Getting a strange error which I'm not able to debug: 
```
KeyError: '`/collections/{collection_name}/exists` not found. Did you mean `/collections/{collection_name}/points`
```

this happens while i try to run the integration tests. I have double checked the actix endpoints and they are correct. but the test fails for some reason.

**Comment by rajdip-b** [comment]
The changes work when i run them on postman, but unfortunately, the tests are failing. Im guessing it has got something to do with how openapi yaml files are configured. To be precise, thse lines:

helpers.py
```python3
operation: APIOperation = SCHEMA[api][method]
```

settings.py
```python3
OPENAPI_FILE = os.environ.get("OPENAPI_FILE", os.path.join(os.path.dirname(ROOT_DIR), '../..', 'openapi-merged.yaml'))
```

**Comment by timvisee** [comment]
@rajdip-b You have updated the OpenAPI definitions, but you still need to generate the final `openapi.json`. That might help with your last mentioned issue.

Simply do:

```bash
./tools/generate_openapi_models.sh
```

This is described in our development guide here: https://github.com/qdrant/qdrant/blob/master/docs/DEVELOPMENT.md

**Comment by rajdip-b** [comment]
Okay gotcha!

File: openapi/tests/openapi_integration/test_collection_exists.py
Comment (rajdip-b) [reply openapi/tests/openapi_integration/test_collection_exists.py:27]: Done!

**Comment by rajdip-b** [comment]
@coszio the tests ran successfully, can you give the pipelines a go?

**Review by timvisee — CHANGES_REQUESTED** [review]
Instead of merging `dev` into here again, could you drop your last commit and rebase on `dev` instead?

**Comment by rajdip-b** [comment]
@timvisee okay you mean reset back to [0ae0254](https://github.com/qdrant/qdrant/pull/3366/commits/0ae025414b92f94ab2f27b87c264cacb66738fde)?

**Comment by rajdip-b** [comment]
can someone tell me why `test-consistency` phase is failing?

**Comment by rajdip-b** [comment]
@timvisee rebase done!

**Comment by rajdip-b** [comment]
> This would be because the generated OpenAPI or gRPC files are outdated. If it fails again it'll list what file to update in the error log.

Kinda strange, since according to the docs, I did add the definitions to the .proto files, added implementations in code, build the code.

**Comment by timvisee** [comment]
> > This would be because the generated OpenAPI or gRPC files are outdated. If it fails again it'll list what file to update in the error log.
> 
> Kinda strange, since according to the docs, I did add the definitions to the .proto files, added implementations in code, build the code.

It complains about the gRPC file again:

https://github.com/qdrant/qdrant/actions/runs/7666406380/job/20894849180?pr=3366#step:9:87

The file might have changed recently on `dev`, which you rebased upon. If you could generate both the OpenAPI and gRPC files one more time, then we should be good.

```bash
$ ./tools/generate_openapi_models.sh
$ ./tools/generate_grpc_docs.sh
```

**Comment by rajdip-b** [comment]
> > > This would be because the generated OpenAPI or gRPC files are outdated. If it fails again it'll list what file to update in the error log.
> > 
> > 
> > Kinda strange, since according to the docs, I did add the definitions to the .proto files, added implementations in code, build the code.
> 
> It complains about the gRPC file again:
> 
> https://github.com/qdrant/qdrant/actions/runs/7666406380/job/20894849180?pr=3366#step:9:87
> 
> The file might have changed recently on `dev`, which you rebased upon. If you could generate both the OpenAPI and gRPC files one more time, then we should be good.
> 
> ```shell
> $ ./tools/generate_openapi_models.sh
> $ ./tools/generate_grpc_docs.sh
> ```

Sure, I'll give it a shot now!

**Comment by rajdip-b** [comment]
@timvisee really clueless about this last test🤨 any ideas what might have went wrong?

File: openapi/openapi-collections.ytt.yaml
Comment (coszio) [inline openapi/openapi-collections.ytt.yaml:145]: This makes the response non-extensible, let's use an object instead. Same way as in the proto spec

**Review by coszio — CHANGES_REQUESTED** [review]

File: openapi/openapi-collections.ytt.yaml
Comment (rajdip-b) [reply openapi/openapi-collections.ytt.yaml:145]: Sure

Comment (rajdip-b) [reply openapi/openapi-collections.ytt.yaml:145]: the response (if 200) is in this format: 
```json
{
    "result": true
}
```

Do you want me to return something else? Or just update the type in the .ytt schema?

Comment (coszio) [reply openapi/openapi-collections.ytt.yaml:145]: more like
```
{
    "result": {
         "exists": true
    }
}
```

Comment (rajdip-b) [reply openapi/openapi-collections.ytt.yaml:145]: ah, cool!

**Comment by rajdip-b** [comment]
@coszio @timvisee some tests in the `test_snapshot.py` file are failing:
```
FAILED openapi_integration/test_snapshot.py::test_full_snapshot_operations[False] - AssertionError: assert 1 == 0
FAILED openapi_integration/test_snapshot.py::test_snapshot_operations_non_wait[False] - AssertionError: assert 1 == 0
FAILED openapi_integration/test_snapshot.py::test_full_snapshot_operations[True] - AssertionError: assert 1 == 0
FAILED openapi_integration/test_snapshot.py::test_snapshot_operations_non_wait[True] - AssertionError: assert 1 == 0
```

I am confused sine this isnt in my scope of change

**Comment by coszio** [comment]
That may be because it is conflicting with some snapshots possibly created outside of the tests. Maybe delete the `storage` and `snapshots` folder before starting the tests?

**Comment by rajdip-b** [comment]
> That may be because it is conflicting with some snapshots possibly created outside of the tests. Maybe delete the `storage` and `snapshots` folder before starting the tests?

Ah okay, so i think it doesn't matter in the workflow since all of them passed. I guess we are clear on this one now?

**Comment by rajdip-b** [comment]
@coszio anything you want me to add more?

File: openapi/tests/openapi_integration/test_collection_exists.py
Comment (timvisee) [inline openapi/tests/openapi_integration/test_collection_exists.py:39]: I wonder... this returns a 404 now. That is of course the expected REST response, but by this logic you could also simply use `GET /collections/something -> 404`.

I wonder if we should return a proper object here as well with HTTP 200 that has `exists: false`.

@coszio What do you think?

Comment (coszio) [reply openapi/tests/openapi_integration/test_collection_exists.py:39]: Yes, I think the whole point of this API is to make it easier to create functions in clients without relying on `404` error response.

**Comment by coszio** [comment]
@rajdip-b FYI, the other attempt (#3472) has just been merged. Thanks a lot for the valuable effort here 🙌

**Comment by rajdip-b** [comment]
> @rajdip-b FYI, the other attempt (#3472) has just been merged. Thanks a lot for the valuable effort here 🙌

Cool! I'll close this PR then.

**Pull request closed without merging**
